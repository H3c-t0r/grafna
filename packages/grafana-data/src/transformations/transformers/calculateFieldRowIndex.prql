# TODO: pick the table at the top of the pipeline
from A
# Row index
derive const = 0
derive `Row` = (row_number const) / (count const)
select {!const}
# Row percentile
derive const = 0
# TODO: Rename this if `Row` is already taken
derive `Row` = (row_number const) / (count const) #
select !{const}
