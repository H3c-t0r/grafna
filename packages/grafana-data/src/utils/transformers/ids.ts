export enum DataTransformerID {
  //  join = 'join', // Pick a field and merge all series based on that field
  append = 'append', // Merge all series together
  //  rotate = 'rotate', // Columns to rows
  reduce = 'reduce', // Run calculations on fields
  filter = 'filter', // Pick some fields
}
