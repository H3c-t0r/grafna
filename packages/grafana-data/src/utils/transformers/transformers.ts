import { DataFrame } from '../../types/dataFrame';
import { Registry, RegistryItemWithOptions } from '../registry';

export type DataTransformer = (data: DataFrame[]) => DataFrame[];

export interface DataTransformerInfo<TOptions = any> extends RegistryItemWithOptions {
  transformer: (options: TOptions) => DataTransformer;
}

export interface DataTransformerConfig<TOptions = any> {
  id: string;
  options: TOptions;
}

// Transformer that does nothing
export const NoopDataTransformer = (data: DataFrame[]) => data;

/**
 * Apply configured transformations to the input data
 */
export function transformDataFrame(options: DataTransformerConfig[], data: DataFrame[]): DataFrame[] {
  let processed = data;
  for (const config of options) {
    const info = dataTransformers.get(config.id);
    const transformer = info.transformer(config.options);
    const after = transformer(processed);

    // Add a key to the metadata if the data changed
    if (after && after !== processed) {
      for (const series of after) {
        if (!series.meta) {
          series.meta = {};
        }
        if (!series.meta.transformations) {
          series.meta.transformations = [info.id];
        } else {
          series.meta.transformations = [...series.meta.transformations, info.id];
        }
      }
      processed = after;
    }
  }
  return processed;
}

// Initalize the Registry

import { filterTransformer, FilterOptions } from './filter';
import { appendTransformer, AppendOptions } from './append';
import { reduceTransformer, ReduceOptions } from './reduce';

class TransformerRegisry extends Registry<DataTransformerInfo> {
  // ------------------------------------------------------------
  // Helper functions for calling directly from javascript code
  // ------------------------------------------------------------

  append(data: DataFrame[], options?: AppendOptions): DataFrame | undefined {
    return appendTransformer.transformer(options || appendTransformer.defaultOptions)(data)[0];
  }

  filter(data: DataFrame[], options: FilterOptions): DataFrame[] {
    return filterTransformer.transformer(options)(data);
  }

  reduce(data: DataFrame[], options: ReduceOptions): DataFrame[] {
    return reduceTransformer.transformer(options)(data);
  }
}

export const dataTransformers = new TransformerRegisry(() => [filterTransformer, appendTransformer, reduceTransformer]);
