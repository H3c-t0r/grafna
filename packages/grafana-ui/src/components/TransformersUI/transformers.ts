import { Registry, RegistryItem } from '@grafana/data';
import { reduceTransformRegistryItem } from './ReduceTransformerEditor';
import { filterFieldsByNameTransformRegistryItem } from './FilterByNameTransformerEditor';

interface TransformerUIRegistyItem extends RegistryItem {
  component: React.ComponentType<any>;
}

export const transformersUIRegistry = new Registry<TransformerUIRegistyItem>(() => {
  return [reduceTransformRegistryItem, filterFieldsByNameTransformRegistryItem];
});
