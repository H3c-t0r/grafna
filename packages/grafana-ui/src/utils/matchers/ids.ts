// This needs to be in its own file to avoid circular references

// A list of some (but not all) matcher IDs
export enum SeriesDataMatcherID {
  // Field Type
  numericFields = 'numericFields',
  timeFields = 'timeFields',
  fieldType = 'fieldType',

  // builtin predicates
  anyMatch = 'anyMatch',
  allMatch = 'allMatch',
  invertMatch = 'invertMatch',
  alwaysMatch = 'alwaysMatch',
  neverMatch = 'neverMatch',
}
