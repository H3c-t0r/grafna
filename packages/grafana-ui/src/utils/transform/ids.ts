export enum SeriesTransformerID {
  join = 'join', // Pick a field and merge all series based on that field
  append = 'append', // Merge all series together
  rotate = 'rotate', // Columns to rows
  calc = 'calc', // Run calculations on fields
  filter = 'filter', // Pick some fields

  // Future
  extract = 'extract', // regex within field to create new fields
  math = 'match', // Do math on the fields
}
