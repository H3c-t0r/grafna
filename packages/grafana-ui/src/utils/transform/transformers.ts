import { SeriesData } from '../../types/data';
import { DataQueryRequest } from '../../types/index';
import { Extension, ExtensionRegistry } from '../extensions';

// A list of some (but not all) transformers
export enum SeriesTransformerID {
  join = 'join', // Pick a field and merge all series based on that field
  append = 'append', // Merge all series together
  rotate = 'rotate', // Columns to rows
  calc = 'calc', // Run calculations on fields
  filter = 'filter', // Pick some fields

  // Future
  extract = 'extract', // regex within field to create new fields
  math = 'match', // Do math on the fields
}

export interface SeriesTransformer<TOptions = any> extends Extension<TOptions> {
  /**
   * Return a modified copy of the series.  If the transform is not or should not
   * be applied, just return the input series
   */
  transform: (options: TOptions, series: SeriesData[], request?: DataQueryRequest) => SeriesData[];
}

export interface SeriesTransformerConfig<TOptions = any> {
  id: string;
  options: TOptions;
}

/**
 * Apply configured transformations to the input data
 */
export function transformSeriesData(
  options: SeriesTransformerConfig[],
  series: SeriesData[],
  request?: DataQueryRequest
): SeriesData[] {
  let processed = series;
  for (const config of options) {
    const transformer = seriesTransformers.get(config.id);
    const after = transformer.transform(config.options, processed);

    // Add a key to the metadata if the data changed
    if (after && after !== processed) {
      for (const series of after) {
        if (!series.meta) {
          series.meta = {};
        }
        if (!series.meta.transformations) {
          series.meta.transformations = [transformer.id];
        } else {
          series.meta.transformations = [...series.meta.transformations, transformer.id];
        }
      }
      processed = after;
    }
  }
  return processed;
}

export const seriesTransformers = new ExtensionRegistry<SeriesTransformer>();

import './filter';
import './calc';
import './append';
