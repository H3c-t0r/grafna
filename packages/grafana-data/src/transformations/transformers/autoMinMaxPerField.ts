import { DataTransformerID } from './ids';
import { DataTransformerInfo } from '../../types/transformations';
import { DataFrame, Field, FieldType, FieldConfig } from '../../types/dataFrame';
import { ReducerID, reduceField } from '../../transformations';
import isNumber from 'lodash/isNumber';

export interface AutoMinMaxPerFieldTransformerOptions {}

export const autoMinMaxPerFieldTransformer: DataTransformerInfo<AutoMinMaxPerFieldTransformerOptions> = {
  id: DataTransformerID.autoMinMaxPerField,
  name: 'Calculate min/max per field',
  description: `Calculate min/max for each field rather than for the complete set of data.
                Useful for cells in bar gauges.`,
  defaultOptions: {},
  transformer: (options: AutoMinMaxPerFieldTransformerOptions) => {
    const reducers = [ReducerID.min, ReducerID.max];
    return (data: DataFrame[]) => {
      console.log(data);
      return data.map((frame, _) => {
        const fields: Field[] = frame.fields.map(field => {
          if (field.type === FieldType.number && field.values) {
            console.log(field);
            console.log(field.config);
            const config: FieldConfig = { ...field.config };
            console.log(config);
            if (!isNumber(config.min) || !isNumber(config.max)) {
              const stats = reduceField({ field, reducers });
              config.min = stats[ReducerID.min];
              config.max = stats[ReducerID.max];
            }
            return {
              ...field,
              config,
            };
          } else {
            return field;
          }
        });
        return {
          ...frame,
          fields,
        };
      });
    };
  },
};
