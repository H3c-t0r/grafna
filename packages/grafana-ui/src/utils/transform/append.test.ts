import { transformSeriesData, seriesTransformers } from './transformers';
import { SeriesTransformerID } from './ids';

const seriesAB = {
  fields: [{ name: 'A' }, { name: 'B' }],
  rows: [
    [1, 2], // A,B
    [1, 2], // A,B
  ],
};

const seriesBC = {
  fields: [{ name: 'B' }, { name: 'C' }],
  rows: [
    [2, 3], // A,B
    [2, 3], // A,B
  ],
};

describe('Append Transformer', () => {
  it('filters by include', () => {
    const cfg = {
      id: SeriesTransformerID.append,
      options: {},
    };
    const x = seriesTransformers.get(SeriesTransformerID.append);
    expect(x.id).toBe(cfg.id);

    const filtered = transformSeriesData([cfg], [seriesAB, seriesBC])[0];
    expect(filtered.fields.length).toBe(3);
    expect(filtered).toMatchSnapshot();
  });
});
