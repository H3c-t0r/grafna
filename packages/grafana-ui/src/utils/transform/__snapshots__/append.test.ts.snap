// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Append Transformer filters by include 1`] = `
Object {
  "fields": Array [
    Object {
      "name": "A",
    },
    Object {
      "name": "B",
    },
    Object {
      "name": "C",
    },
  ],
  "meta": Object {
    "transformations": Array [
      "append",
    ],
  },
  "rows": Array [
    Array [
      1,
      2,
    ],
    Array [
      1,
      2,
    ],
    Array [
      undefined,
      2,
      3,
    ],
    Array [
      undefined,
      2,
      3,
    ],
  ],
}
`;
